aws_lambda_powertools/utilities/data_classes/common.py

```diff
@@ -28,6 +28,49 @@ def __iter__(self) -> Iterator:
     def __len__(self) -> int:
         return len(self._data)
 
+    def __str__(self) -> str:
+        return str(self._str_helper())
+
+    def _str_helper(self) -> Dict[str, Any]:
+        """
+        Recursively get a Dictionary of DictWrapper properties primarily
+        for use by __str__ for debugging purposes.
+
+        Will remove "raw_event" properties, and any defined by the Data Class
+        `_str_sensitive_data` list field.
+        This should be used in case where secrets, such as access keys, are
+        stored in the Data Class but should not be logged out.
+        """
+        properties = self._properties()
+        sensitive_properties = ["raw_event"]
+        try:
+            sensitive_properties.extend(self.__getattribute__("_str_sensitive_data"))
+        except AttributeError:
+            pass
+
+        result: Dict[str, Any] = {}
+        for property_key in properties:
+            if property_key in sensitive_properties:
+                result[property_key] = "[SENSITIVE]"
+            else:
+                try:
+                    property_value = getattr(self, property_key)
+                    result[property_key] = property_value
+
+                    if issubclass(property_value.__class__, DictWrapper):
+                        result[property_key] = property_value._str_helper()
+                    elif isinstance(property_value, list):
+                        for seq, item in enumerate(property_value):
+                            if issubclass(item.__class__, DictWrapper):
+                                result[property_key][seq] = item._str_helper()
+                except Exception as e:
+                    result[property_key] = f"[EXCEPTION {type(e)}]"
```

Comment:
I agree that presenting an exception could make a developer/user think something is wrong.
I had weighed up discarding the property, however felt hiding information was worse in a troubleshooting situation.

This appears to be most likely when some user supplied event data could be either plaintext, or JSON.
I figured that if anyone saw a `json.JSONDecodeError` would hopefully see the corresponding non-JSON property is plain text.
We could specifically handle JSON, and return "[Cannot Decode as JSON]".

Reply:
How about using `None`? It's hard to check all the exceptions that can happen, it could be JSON, it could be another class... so using `None` is more generalist.

{'account_id': '111111111111', 'data': {'action_configuration': {'configuration': {**'decoded_user_parameters': 'None'**, 'function_name': 'MyLambdaFunctionForAWSCodePipeline', 'raw_event': '[SENSITIVE]', 'user_parameters': 'some-input-such-as-a-URL'}, 'raw_event': '[SENSITIVE]'}, 'artifact_credentials': {'access_key_id': 'AKIAIOSFODNN7EXAMPLE', 'expiration_time': None, 'raw_event': '[SENSITIVE]', 'secret_access_key': '[SENSITIVE]', 'session_token': '[SENSITIVE]'}, 'continuation_token': 'A continuation token if continuing job', 'encryption_key': None, 'input_artifacts': [{'location': {'get_type': 'S3', 'raw_event': '[SENSITIVE]', 's3_location': {'bucket_name': 'the name of the bucket configured as the pipeline artifact store in Amazon S3, for example codepipeline-us-east-2-1234567890', 'key': 'the name of the application, for example CodePipelineDemoApplication.zip', 'object_key': 'the name of the application, for example CodePipelineDemoApplication.zip', 'raw_event': '[SENSITIVE]'}}, 'name': 'ArtifactName', 'raw_event': '[SENSITIVE]', 'revision': None}], 'output_artifacts': [], 'raw_event': '[SENSITIVE]'}, 'decoded_user_parameters': 'None', 'get_id': '11111111-abcd-1111-abcd-111111abcdef', 'input_bucket_name': 'the name of the bucket configured as the pipeline artifact store in Amazon S3, for example codepipeline-us-east-2-1234567890', 'input_object_key': 'the name of the application, for example CodePipelineDemoApplication.zip', 'raw_event': '[SENSITIVE]', 'user_parameters': 'some-input-such-as-a-URL'}

Reply:
Yes `None` would work, but this would separate perspection from reality (i.e. calling the actual property will raise an exception instead of returning `None`).
This could result in the user believing the propery has exception handling, and attempt to do a comparison
```python
if broken_property_json() is None:  # <-- raises unexpected Exception
   # do something
```

I'd rather potentially a user be confused and raise the question "did my code fail?", and for them to investigate, rather than provide them false information (i.e. that calling that property would return `None`).

Reply:
Hi @neilramsay sorry for the late reply! Yes, I agree with you that we can go to the same error ..

I changed the code and the non-deserialized fields will have the string value `[Cannot be deserialized]`. I also created documentation for this, so I think we are sure that the reason for this value will be clear to the user.

Reply:
Yes, I agree this is a pragmatic approach, and increases clarity.
I believe this has resolved the conversation topic, so going to resolve.